File: insight-lang/src/main/java/com/github/lonelylockley/archinsight/lexer/IndentHelper.java
package com.github.lonelylockley.archinsight.lexer;

import com.github.lonelylockley.insight.lang.InsightLexer;
import org.antlr.v4.runtime.CommonToken;
import org.antlr.v4.runtime.RecognitionException;
import org.antlr.v4.runtime.Token;

import java.util.LinkedList;

import static org.antlr.v4.runtime.Lexer.DEFAULT_TOKEN_CHANNEL;

public class IndentHelper {

    public static final int INDENT_LENGTH = 4;

    private final InsightLexer lexer;
    private final LinkedList<Token> waitlist;

    private int indentation = 0;
    private boolean wrapped = false;
    private boolean singleLineMode = false;
    private LexerState state;
    private boolean eofReached = false;

    public IndentHelper(InsightLexer lexer) {
        this.lexer = lexer;
        this.waitlist = new LinkedList<>();
        this.state = new LexerState();
    }

    private String stripNewlineCharacters(String src) {
        return src.replaceAll("[\n\r]", "");
    }

    private int countNewLines(String src) {
        var lastIndex = 0;
        var count = 0;
        while(lastIndex != -1){
            lastIndex = src.indexOf('\n', lastIndex);
            if(lastIndex != -1){
                count++;
                lastIndex += 1;
            }
        }
        return count;
    }

    private String indentationError(int actual, int expected) {
        return String.format("line %d:%d incorrect indentation. current position: %d, expected: %d", lexer.getLine(), lexer.getCharPositionInLine(), actual, expected);
    }

    private CommonToken createToken(int type, String text, int length, int lineCorrection, int offsetCorrection) {
        var stop = lexer.getCharIndex() + offsetCorrection;
        var start = stop - length + 1;
        var tkn = new CommonToken(lexer.getTokenFactorySourcePair(), type, DEFAULT_TOKEN_CHANNEL, start, stop);
        tkn.setText(text);
        tkn.setLine(lexer.getLine() + lineCorrection);
        return tkn;
    }

    private int calculateLengthCorrection() {
        return lexer.getText() == null ? 0 : -lexer.getText().length();
    }

    private int calculateIndentation(String indent) {
        int count = 0;
        for (char ch : indent.toCharArray()) {
            if (ch == '\t') {
                count += INDENT_LENGTH;
            }
            else {
                count++;
            }
        }
        if (count % INDENT_LENGTH != 0) {
            lexer.getErrorListenerDispatch().syntaxError(lexer, lexer.getToken(), lexer._tokenStartLine, lexer.getCharPositionInLine(), indentationError(lexer.getCharPositionInLine(), indentation * INDENT_LENGTH),
                new RecognitionException("incorrect indentation", lexer, lexer._input, null)
            );
        }
        return count / INDENT_LENGTH;
    }

    private void fireIndents(int desiredIndentation, int linesCorrection) {
        while (indentation < desiredIndentation) {
            indentation++;
            state.incIndentation();
            waitlist.add(createToken(InsightLexer.INDENT, "<INDENT>", INDENT_LENGTH, 0, -linesCorrection));
        }
    }

    private void fireDedents(int desiredIndentation, int linesCorrection) {
        while (indentation > desiredIndentation) {
            waitlist.add(createToken(InsightLexer.DEDENT, "<DEDENT>", 0, -linesCorrection, 0));
            indentation--;
            state.decIndentation();
        }
    }

    public void checkIndentation() {
        var newLines = countNewLines(lexer.getText());
        var newIndentation = calculateIndentation(stripNewlineCharacters(lexer.getText()));
        waitlist.add(createToken(InsightLexer.EOL, "\n", 1, -newLines, calculateLengthCorrection()));
        if (newIndentation > indentation) {
            fireIndents(newIndentation, newLines);
        }
        else
        if (!singleLineMode) {
            fireDedents(newIndentation, newLines);
        }
    }

    public void wrapValue() {
        if (!wrapped) {
            wrapped = true;
            state.setWasText();
            waitlist.add(createToken(InsightLexer.WRAP, "<WRAP>", 0, 0, calculateLengthCorrection()));
        }
    }

    public void unwrapValue() {
        var newLines = countNewLines(lexer.getText());
        var newIndentation = calculateIndentation(stripNewlineCharacters(lexer.getText()));
        if (newIndentation == indentation + 1) {
            waitlist.add(createToken(InsightLexer.TEXT, "\n", 1, -newLines, calculateLengthCorrection()));
        }
        else
        if ((newIndentation <= indentation) && wrapped) {
            wrapped = false;
            if (!singleLineMode) {
                state.resetWasText();
                waitlist.add(createToken(InsightLexer.UNWRAP, "<UNWRAP>", 0, -newLines, calculateLengthCorrection()));
                waitlist.add(createToken(InsightLexer.EOL, "\n", 1, -newLines, calculateLengthCorrection()));
                lexer.popMode();
                fireDedents(newIndentation, newLines);
            }
        }
        else {
            lexer.getErrorListenerDispatch().syntaxError(lexer, lexer.getToken(), lexer._tokenStartLine, lexer.getCharPositionInLine(), indentationError(lexer.getCharPositionInLine(), indentation * INDENT_LENGTH),
                new RecognitionException("incorrect indentation", lexer, lexer._input, null)
            );
        }
    }

    public void processEOF(Token eof) {
        if (wrapped) {
            //waitlist.add(createToken(InsightLexer.EOL, "\n", 1, 0, 0));
            lexer.setText("\n");
            unwrapValue();
        }
        else
        if (lexer.getInputStream().LA(-1) != 10 && lexer.getText() == null) {
            //waitlist.add(createToken(InsightLexer.EOL, "\n", 1, 0, 0));
            lexer.setText("\n");
            checkIndentation();
        }
        waitlist.add(eof);
    }

    public Token nextToken() {
        Token tkn;
        if (!waitlist.isEmpty()) {
            tkn = waitlist.pollFirst();
        }
        else {
            tkn = lexer.supplyToken();
            if (!waitlist.isEmpty()) {
                waitlist.add(tkn);
                tkn = waitlist.pollFirst();
            }
        }
        if (tkn.getType() == Token.EOF && !eofReached) {
            eofReached = true;
            processEOF(tkn);
            if (!waitlist.isEmpty()) {
                tkn = waitlist.pollFirst();
            }
        }
        final String rawType = lexer.getVocabulary().getSymbolicName(tkn.getType());
        System.out.println("---- " + rawType + " [line=" + tkn.getLine() + ",from=" + tkn.getStartIndex() + ",to=" + tkn.getStopIndex() + ",mode=" + lexer._mode + ",channel=" + tkn.getChannel() + "] = `" + tkn.getText() + "`");
        return tkn;
    }

    public void enableSingleLineMode() {
        singleLineMode = true;
    }

    public LexerState snapshotState() {
        return (LexerState) state.clone();
    }

    public void restoreState(LexerState state) {
        //Token tkn = tokenSupplier.get();
        lexer.reset();
        this.state = (LexerState) state.clone();
        indentation = state.getIndentation();
        if (state.wasText()) {
            wrapped = true;
            lexer.pushMode(InsightLexer.VALUE_MODE);
        }
    }

}

File: editor-ui/frontend/src/IndentHelper.ts
import { CommonToken, Token } from 'antlr4ng';
import LexerState from './LexerState';
import { InsightLexer } from '../generated/insight-lang/InsightLexer';

class IndentationException extends Error {

    constructor(msg: string) {
        super(msg);
    }

    public static createErrorMsg(lexer: InsightLexer, actual: number, expected: number): string {
        return "line " + lexer.line + ":" + lexer.column + " incorrect indentation. current position: " + actual + ", expected: " + expected;
    }

}

class IndentHelper {

    INDENT_LENGTH: number = 4;

    private lexer: InsightLexer;
    private waitlist: Token[];

    private indentation: number = 0;
    private wrapped: boolean = false;
    private singleLineMode: boolean = false;
    private state: LexerState;
    private tokenId: number = 0;

    constructor(lexer: InsightLexer) {
        this.lexer = lexer;
        this.waitlist = [];
        this.state = new LexerState();
    }

    private stripNewlineCharacters(src: string | undefined): string {
        return src == undefined ? "" : src.replace(/\r?\n/g, "");
    }

    private countNewLines(src: string): number {
        let lastIndex = 0;
        let count = 0;
        while(lastIndex != -1){
            lastIndex = src.indexOf('\n', lastIndex);
            if(lastIndex != -1){
                count++;
                lastIndex += 1;
            }
        }
        return count;
    }

    private calculateLengthCorrection(): number {
        return this.lexer.text == null ? 0 : -this.lexer.text.length;
    }

    private createToken(type: number, text: string, tokenLength: number, lineCorrection = 0, offsetCorrection = 0, column: number | undefined = undefined): CommonToken {
        let stop = this.lexer.getCharIndex() + offsetCorrection;
        let start = stop - tokenLength + 1;
        let tkn = CommonToken.fromSource([this.lexer, this.lexer.inputStream], type, InsightLexer.DEFAULT_TOKEN_CHANNEL, start, stop);
        tkn.setLine(this.lexer.line + lineCorrection);
        tkn.setText(text);
        if (column != undefined) {
            tkn.setCharPositionInLine(column);
        }
        return tkn;
    }

    private calculateIndentation(indent: string): number {
        let count = 0;
        for (const ch of indent) {
            if (ch == '\t') {
                count += this.INDENT_LENGTH;
            }
            else {
                count++;
            }
        }
        if (count % this.INDENT_LENGTH != 0) {
            this.lexer.errorListenerDispatch.syntaxError?.(
                this.lexer,
                this.createToken(InsightLexer.INDENT, " ".repeat(this.lexer.column), this.lexer.column, 0, 0, 0),
                this.lexer.line,
                0,
                IndentationException.createErrorMsg(this.lexer, this.lexer.column, this.indentation * this.INDENT_LENGTH),
                null
            );
        }
        return count / this.INDENT_LENGTH;
    }


    private fireIndents(desiredIndentation: number, linesCorrection: number) {
        while (this.indentation < desiredIndentation) {
            this.indentation++;
            this.state.incIndentation();
            this.waitlist.push(this.createToken(InsightLexer.INDENT, "<INDENT>", this.INDENT_LENGTH, 0, -linesCorrection));
        }
    }

    private fireDedents(desiredIndentation: number, linesCorrection: number) {
        while (this.indentation > desiredIndentation) {
            this.waitlist.push(this.createToken(InsightLexer.DEDENT, "<DEDENT>", 0, -linesCorrection, 0));
            this.indentation--;
            this.state.decIndentation();
        }
    }

    public checkIndentation() {
        let newLines = this.countNewLines(this.lexer.text);
        let newIndentation = this.calculateIndentation(this.stripNewlineCharacters(this.lexer.text));
        this.waitlist.push(this.createToken(InsightLexer.EOL, "\n", 1, -newLines, this.calculateLengthCorrection()));
        if (newIndentation > this.indentation) {
            this.fireIndents(newIndentation, newLines);
        }
        else {
            if (!this.singleLineMode) {
                this.fireDedents(newIndentation, newLines);
            }
        }
    }

    public wrapValue() {
        if (!this.wrapped) {
            this.wrapped = true;
            this.state.setWasText();
            this.waitlist.push(this.createToken(InsightLexer.WRAP, "<WRAP>", 0, 0, this.calculateLengthCorrection()));
        }
    }

    public unwrapValue() {
        let newLines = this.countNewLines(this.lexer.text);
        let newIndentation = this.calculateIndentation(this.stripNewlineCharacters(this.lexer.text));
        if (newIndentation == this.indentation + 1) {
            this.waitlist.push(this.createToken(InsightLexer.TEXT, "\n", 1, -newLines, this.calculateLengthCorrection()));
        }
        else
        if ((newIndentation <= this.indentation) && this.wrapped) {
            this.wrapped = false;
            if (!(this.singleLineMode && this.lexer.inputStream.LA(1) == -1)) {
                this.state.resetWasText();
                this.waitlist.push(this.createToken(InsightLexer.UNWRAP, "<UNWRAP>", 0, -newLines, this.calculateLengthCorrection()));
                this.waitlist.push(this.createToken(InsightLexer.EOL, "\n", 1, -newLines,  this.calculateLengthCorrection()));
                this.lexer.popMode();
                this.fireDedents(newIndentation, newLines);
            }
        }
        else {
            this.lexer.errorListenerDispatch.syntaxError?.(
                this.lexer,
                this.createToken(InsightLexer.INDENT, " ".repeat(this.lexer.column), this.lexer.column, 0, 0, 0),
                this.lexer.line,
                0,
                IndentationException.createErrorMsg(this.lexer, this.lexer.column, this.indentation * this.INDENT_LENGTH),
                null
            );
        }
    }

    public processEOF(eof: Token) {
        if (this.wrapped) {
            this.lexer.text = "\n";
            this.unwrapValue();
            this.waitlist.push(eof);
        }
        else
        if (this.lexer.inputStream.LA(-1) != 10 && this.lexer.text == undefined) {
            this.lexer.text = "\n";
            this.checkIndentation();
            this.waitlist.push(eof);
        }
        else
        if (!this.singleLineMode && !(this.waitlist.length > 0 && this.waitlist[this.waitlist.length - 1].type == InsightLexer.EOF)) {
            this.fireDedents(0, 0);
            this.waitlist.push(eof);
        }
    }

    public nextToken(): Token {
        let tkn: Token;
        if (this.waitlist.length > 0) {
            tkn = this.waitlist.shift()!;
        }
        else {
            tkn = this.lexer.supplyToken();
            if (this.waitlist.length > 0) {
                this.waitlist.push(tkn);
                tkn = this.waitlist.shift()!;
            }
        }
        if (tkn.type == Token.EOF) {
            this.processEOF(tkn);
            if (this.waitlist.length > 0) {
                this.waitlist.push(tkn);
                tkn =this.waitlist.shift()!;
            }
        }
        tkn.tokenIndex = this.tokenId;
        this.tokenId++;
        // const rawType = this.lexer.vocabulary.getSymbolicName(tkn.type);
        // console.log("---- " + rawType + " [idx=" + tkn.tokenIndex + "line=" + tkn.line + ",from=" + tkn.start + ",to=" + tkn.stop + ",mode=" + this.lexer.mode + ",channel=" + tkn.channel + "] = `" + tkn.text + "`");
        return tkn;
    }

    public enableSingleLineMode() {
        this.singleLineMode = true;
    }

    public snapshotState(): LexerState {
        return this.state.clone() as LexerState;
    }

    public restoreState(state: LexerState) {
        this.lexer.reset();
        this.state = state.clone() as LexerState;
        this.indentation = state.getIndentation();
        if (state.wasText()) {
            this.wrapped = true;
            this.lexer.pushMode(InsightLexer.VALUE_MODE);
        }
    }

}

export default IndentHelper;
